### ISSUE TITLE
Photo ID is old 10-integer version

### ISSUE THREAD

**Issue by bennettscience** [LycheeOrg/Lychee-Laravel#72]
Using a new Laravel installation with photos uploaded via UI. Thumbnails display fine in Unsorted, but clicking a single image breaks the UI. The logs show that the specified file is missing. Here's the log output and mysql table after uploading a single file:

Logs:

```
    2019-02-10 18:02:43 -- error -- App\Http\Controllers\PhotoController::get -- 50 -- Could not find specified photo
    2019-02-10 18:02:43 -- error -- App\Configs::get_value -- 126 -- compression_quality does not exist in config !
    2019-02-10 18:02:42 -- notice -- App\Image\ImagickHandler::scale -- 47 -- Saving thumb to /var/www/lychee-laravel/public/uploads/small/8010f292035b61cb8b413b1b9ff4ffe0.jpg
    2019-02-10 18:02:41 -- notice -- App\ModelFunctions\PhotoFunctions::createMedium -- 166 -- No resize (image is too small)!
    2019-02-10 18:02:41 -- notice -- App\ModelFunctions\PhotoFunctions::add -- 389 -- 8010f292035b61cb8b413b1b9ff4ffe0.jpeg
    2019-02-10 18:02:41 -- notice -- App\Image\ImagickHandler::crop -- 82 -- Saving thumb to /var/www/lychee-laravel/public/uploads/thumb/8010f292035b61cb8b413b1b9ff4ffe0@2x.jpeg
    2019-02-10 18:02:41 -- notice -- App\Image\ImagickHandler::crop -- 82 -- Saving thumb to /var/www/lychee-laravel/public/uploads/thumb/8010f292035b61cb8b413b1b9ff4ffe0.jpeg
    2019-02-10 18:02:41 -- notice -- App\ModelFunctions\PhotoFunctions::createThumb -- 115 -- Photo URL is 8010f292035b61cb8b413b1b9ff4ffe0.jpg
```

MySQL table:
```
 id             | title                     | description | url                                  | tags | public | owner_id | type       | width | height | size    | iso | aperture | make | model | lens | shutter | focal | latitude | longitude | altitude | takestamp | star | thumbUrl                              | album_id | checksum                                 | medium | small | license | created_at          | updated_at          |
+----------------+---------------------------+-------------+--------------------------------------+------+--------+----------+------------+-------+--------+---------+-----+----------+------+-------+------+---------+-------+----------+-----------+----------+-----------+------+---------------------------------------+----------+------------------------------------------+--------+-------+---------+---------------------+---------------------+
| 15498217614429 | the-gallery_28592155458_o |             | 8010f292035b61cb8b413b1b9ff4ffe0.jpg |      |      0 |        0 | image/jpeg |   600 |    800 | 96.1 KB |     |          |      |       |      |         |       |     NULL |      NULL |     NULL | NULL      |    0 | 8010f292035b61cb8b413b1b9ff4ffe0.jpeg |     NULL | e60856a3915762c18d4552387a9a6f2409c87019 |      0 |     1 | none    | 2019-02-10 18:02:42 | 2019-02-10 18:02:42 |
```

Looking at `album.json.photos` I see a single `photo` object (expected) but `photo.id` does not match the MySQL ID:

`album.json.photo[0]`:
```
album: null
altitude: null
aperture: ""
cameraDate: "0"
description: ""
focal: ""
height: 800
id: 2147483647 // Should this be 15498217614429?
iso: ""
latitude: null
lens: ""
license: "none"
longitude: null
make: ""
medium: ""
model: ""
nextPhoto: ""
previousPhoto: ""
public: "0"
shutter: ""
size: "96.1 KB"
small: "uploads/small/8010f292035b61cb8b413b1b9ff4ffe0.jpg"
star: "0"
sysdate: "10 February 2019"
tags: ""
takedate: ""
thumbUrl: "uploads/thumb/8010f292035b61cb8b413b1b9ff4ffe0.jpeg"
title: "the-gallery_28592155458_o"
type: "image/jpeg"
url: "uploads/big/8010f292035b61cb8b413b1b9ff4ffe0.jpg"
width: 600
```

The browser throws a console error because it can't find the correct photo object to display:
Console error:
```
main.js:1497 Uncaught TypeError: Cannot read property 'split' of undefined
    at Object.build.tags (main.js:1497)
    at Object.sidebar.createStructure.photo (main.js:5191)
    at Object.<anonymous> (main.js:6535)
    at Object.sidebar (main.js:6525)
    at Object.init (main.js:6360)
    at main.js:3834
    at Object.success (main.js:1193)
    at u (main.js:2)
    at Object.fireWith [as resolveWith] (main.js:2)
    at k (main.js:2)
```

Laravel and Lychee-front were both merged today.

**Comment by ildyria** [comment]
I can't reproduce that. Surprising.

**Comment by ildyria** [comment]
> id: 2147483647 // Should this be 15498217614429? 

Yes.

**Comment by d7415** [comment]
> > id: 2147483647 // Should this be 15498217614429?

That's 10 chars... Coincidence, or could it to be with the old ID generation?

**Comment by ildyria** [comment]
I think this is linked yes

**Comment by bennettscience** [comment]
I'm going to try removing the entire install again to see if I can reproduce it. I did a full wipe of `uploads/*` and dumped the tables with no change.

**Comment by bennettscience** [comment]
Was there a script at some point to regenerate IDs if they're the old version?

**Comment by ildyria** [comment]
There was one. It was removed onced we switched to `BigIncrement` instead of `Increment` as the numbers were now big enough to be stored.

**Comment by bennettscience** [comment]
Ok, I'll have to do more debugging on this one. I finished fixing #75 and moved photos over to the new install, but this error persisted because the migrated IDs don't match.

**Comment by bennettscience** [comment]
I did a fresh install tonight and then uploaded a photo. The thumbnail updates and a MySQL entry is created. `album.json.photos` is still showing the wrong ID number for the photo, which points the browser to the wrong URL.

Generated URL:
`http://192.168.2.150:8080/#r/2147483647`

`album.json.photos`:
```
photos: Array(1)
0:
album: null
altitude: null
aperture: ""
cameraDate: "0"
description: ""
focal: ""
height: 573
id: 2147483647
iso: ""
latitude: null
lens: ""
license: "none"
longitude: null
make: ""
medium: ""
model: ""
nextPhoto: ""
previousPhoto: ""
public: "0"
shutter: ""
size: "602.8 KB"
small: "uploads/small/92234ce5c88289cbddd0f562e2e5a931.png"
star: "0"
sysdate: "21 February 2019"
tags: ""
takedate: ""
thumbUrl: "uploads/thumb/92234ce5c88289cbddd0f562e2e5a931.jpeg"
title: "pdOnAir"
type: "image/png"
url: "uploads/big/92234ce5c88289cbddd0f562e2e5a931.png"
width: 1024
```
The MySQL entry has the correct ID stored in the table:
```
| id             | title   | description | url                                  | tags | public | owner_id | type      | width | height | size     | iso | aperture | make | model | lens | shutter | focal | latitude | longitude | altitude | takestamp | star | thumbUrl                              | album_id | checksum                                 | medium | small | license | created_at          | updated_at          |
| 15507159841546 | pdOnAir |             | 92234ce5c88289cbddd0f562e2e5a931.png |      |      0 |        0 | image/png |  1024 |    573 | 602.8 KB |     |          |      |       |      |         |       |     NULL |      NULL |     NULL | NULL      |    0 | 92234ce5c88289cbddd0f562e2e5a931.jpeg |     NULL | 79f664b908ecf58a3cd1f681e04adc57a58d5173 |      0 |     1 | none    | 2019-02-21 02:26:28 | 2019-02-21 02:26:28 |
```

Replacing the generated URL with the correct ID takes me to the full photo as expected.

## What I've tried

- Fresh laravel install
- Migrating old Lychee tables
- Uploading to empty photos table
- Clear laravel cache with `artisan config:cache`
- Restart MySQL server
- Restart Apache server

Any other tips on debugging this one? I can give login access to this account if that'd be helpful as it's for testing only.

**Comment by bennettscience** [comment]
Ok, I did more testing and found something really weird. In the several days since deleting the old instance and installing again, I noticed that the URL generated has the wrong 10-digit ID number for any photo, new or old. 

I've restarted Apache and the MySQL server as well as the entire server itself. 

I've cleared browser cache and used browsers that have never visited the site.

All IDs point to `2147483647`.

An empty MySQL database shows one photo with a functional ID (when I replace the browser URL ID with the database ID, the photo loads as expected).

**MySQL `photos` table**
```
| id             | title   | description | url                                  | tags | public | owner_id | type      | width | height | size     | iso | aperture | make | model | lens | shutter | focal | latitude | longitude | altitude | takestamp | star | thumbUrl                              | album_id | checksum                                 | medium | small | license | created_at          | updated_at          |
| 15507159841546 | pdOnAir |             | 92234ce5c88289cbddd0f562e2e5a931.png |      |      0 |        0 | image/png |  1024 |    573 | 602.8 KB |     |          |      |       |      |         |       |     NULL |      NULL |     NULL | NULL      |    0 | 92234ce5c88289cbddd0f562e2e5a931.jpeg |     NULL | 79f664b908ecf58a3cd1f681e04adc57a58d5173 |      0 |     1 | none    | 2019-02-21 02:26:28 | 2019-02-21 02:26:28 |
```
What else could be storing and returning that ID?

**Comment by d7415** [comment]
> All IDs point to `2147483647`.

32 bit signed int....? *looks*

**Comment by d7415** [comment]
I can't find it, but that makes me think that the value might be being truncated from at 14-digit string or integer into a 32-bit, signed integer that's (silently) maxing out.

Are your server/client/both 32 bit?

**Comment by bennettscience** [comment]
@d7415 Thanks for the idea, I'll try it tonight at home (SSH is blocked at work). It's a Raspberry Pi 3 running the latest Raspbian ARM edition, which is 32 bit. It has a 64 bit processor, but I don't think that would matter...

MySQL is also current, but I have to check the specs at home.

**Comment by bennettscience** [comment]
The RasPi 3 runs an Armv7 architecture, which is 32 bit, which may be the problem. Any ideas on how to fix that?

**Comment by d7415** [comment]
> Are there any methods to work around the architecture limitations?

I don't think so :(
Unless we can implement a different work around (like Lychee's use of strings or @ildyria's previous solution (which I think had its own problems)), we might have to mark 64 bit system+php a requirement of Lychee-Laravel.

**Comment by bennettscience** [comment]
Yeah, let's add it to the wiki. I can't imagine a ton of users are self hosting on a raspberry pi, but this would help prevent headaches for ones who are. [I found a Fedora server image that will run in 64bit](https://fedoraproject.org/wiki/Architectures/ARM/Raspberry_Pi#aarch64_supported_images_for_Raspberry_Pi_3) on a Pi 3, so maybe I'll give that a try. If that works, we can also specify that option on the wiki.

**Comment by bennettscience** [comment]
Fixed with #171.
